## Analytics for Just Dice: Understanding the Data...

In [2]:
# importing necessary libraries
import pandas as pd
import  matplotlib.pyplot as plt
import seaborn as sns
import cufflinks as cf
cf.go_offline()

from IPython.display import HTML

We have the following data - 

    - adspends - Expenditure on advertisments

    - installs - Install related information

    - revenue  - Revenue generated by in-game advertisements

    - payouts  - Payout to the different users for downloading partner-apps

Let us have a quick overview of the available data

In [3]:
# loading data
adspend = pd.read_csv("Data/adspend.csv")
installs = pd.read_csv("Data/installs.csv")
payouts = pd.read_csv("Data/payouts.csv")
revenue = pd.read_csv("Data/revenue.csv")

In [4]:
len_i = len(installs)

In [5]:
html_code_single = """
<div>
      <p style="text-align: center; margin-top: 0;">{text}</p>
      <p style="text-align: center; font-size: 300%; margin-bottom: 0;">{value}</p>
 </div>
"""

In [6]:
# Total installations
display(HTML(html_code_single.format(text = "Total Number of Installations in the year 2022", value = len_i)))

In [7]:
sum_adspend = round(sum(adspend["value_usd"]), 2)

In [8]:
sum_payouts = round(sum(payouts["value_usd"]), 2)

In [9]:
sum_revenue = round(sum(revenue["value_usd"]), 2)

In [10]:
net_profit = round(sum(revenue["value_usd"]) - sum(payouts["value_usd"]) -sum(adspend["value_usd"]), 2)

In [30]:
html_code = """
<div style="display: flex;  left: 0; justify-content: center; gap: 10%; padding: 10px">
  <div style="border: 2px solid black; padding: 10px;">
      <p style="text-align: center; margin-top: 0;">Total Revenue generated </p>
      <p style="font-size: 300%; margin-bottom: 0;">{sum_revenue}</p>
  </div>
  <div style="border: 2px solid black; padding: 10px; margin-right: 10px;">
      <p style="text-align: center; margin-top: 0;">Total expenditure on Advertisments</p>
      <p style="text-align: center; font-size: 300%; margin-bottom: 10px;">{sum_adspend}</p>
  </div>
  <div style="border: 2px solid black; padding: 10px;">
      <p style="text-align: center; margin-top: 0;">Total expenditure on Payouts </p>
      <p style="font-size: 300%; margin-bottom: 0;">{sum_payouts}</p>
  </div>
</div>
"""

display(HTML(html_code.format(sum_adspend=sum_adspend, sum_revenue = sum_revenue, sum_payouts = sum_payouts)))

In [12]:
display(HTML(html_code_single.format(text = "Net Profit", value = net_profit)))

In [13]:
### Trends across Countries

Let us see how our applications perform in different countries

### Trends across Countries

##### Which Countries Are Leading the Spend Race?

In [14]:
adspend_country = adspend.groupby("country_id")["value_usd"].sum().reset_index()
adspend_country.sort_values(by = "value_usd", ascending=False)

country_id      value_usd
0           1  243186.687782
2         109   10053.718011
3         213     501.178006
1          17     334.392000

In [15]:
adspend_country.iplot(kind = "pie", 
                  values = "value_usd", 
                  labels = "country_id",
                  title = "Ad Expenses per Country",
                 )

95% of our ad spends are on Country 1, followed by ~4% on Country 109.

##### Which Countries are Downloading the Most?

In [53]:
installs["country_id"].value_counts().iplot(kind="bar",
                                            layout = {"xaxis": {"type":"category", 
                                                                "title": "Country ID"},
                                                      "yaxis": {"title": "Number of Installs"},
                                                      "title":"Number of Installs per Country"}
                                           )

### Trends across Networks

Let us see which networks are most effective in acquiring users

##### Expenditure on Advertisements Network-wise

In [17]:
adspend_net = adspend.groupby("network_id")["value_usd"].sum().reset_index()
adspend_net.sort_values(by = "value_usd", ascending=False)

network_id      value_usd
1          60  214838.614914
0          10   39237.360885

In [18]:
adspend_net.iplot(kind = "pie", 
                  values = "value_usd", 
                  labels = "network_id",
                  textinfo='percent+label',
                  title = "Ad Expenses per network"
                 )

Looks like we have two major networks where we advertise our applicatons.

85% of the expenditure is on network with identifier 60

##### Install Trends Network-wise

In [152]:
installs["network_id"].value_counts().iplot(kind="bar",
                                            layout = {"xaxis": {"type":"category",
                                                               "title":"Network ID"},
                                                      "yaxis":{"title":"Number of Installs"},
                                                      "title":"Number of Installs per Network"
                                                     }
                                           )

It's not surprising that the majority of our installs come from Network 60, as it has the highest number of ad expenses. 

However, it's worth noting that we also get a significant number of installs from Network 26, despite not spending on advertisements there.

To make the most of our ad budget, we could consider allocating some funds towards Network 26, as it has the potential to yield good results at a lower cost.

##### Ad expenditure per country and network

In [20]:
adspend_country_net = pd.pivot_table(adspend, index='country_id', columns='network_id', values='value_usd', aggfunc='sum')

In [154]:
adspend_country_net.iplot(kind = "bar",
                          layout = {"xaxis": {"type":"category",
                                             "title":"Network ID"},
                                    "yaxis":{"title":"Total Amount Spent"},
                                    "title": "Ad Expenses per Country per Network"},
                         )

Clearly, network 60 dominates in all countries

###  Installations across Countries and Networks

In [22]:
install_country_net = pd.pivot_table(installs, index='country_id', columns='network_id', values='install_id', aggfunc='count')

In [174]:
install_country_net.iplot(kind = "bar",
                          layout = {"xaxis": {"type":"category",
                                             "title":"Country ID",},
                                    "yaxis":{"title":"Number of Installs"},
                                    "legend":{"title":"Network ID",},
                                   "title": "Installs per Country per Network",
                                   },
                         )

As expected, Network 60 is still leading in terms of installs, followed by 60 and 10.

Interestingly, Network 26 shows a considerable impact in country 109.

### The Rich Kids of Advertising

##### Top 10 Apps with maximum Ad expenses 

In [235]:
adspend.groupby("client_id")["value_usd"].sum().\
                                        sort_values(ascending=False).\
                                        head(10).reset_index()

client_id      value_usd
0        174  104606.688250
1         94   71636.985414
2        121   42158.106030
3         71   11021.333005
4        189    4474.769952
5        104    3187.530988
6        256    2968.805070
7        302    2631.670026
8        275    1998.470014
9        380    1764.217032

##### Top 10 Apps with maximum Installs

In [238]:
installs.groupby("app_id")["install_id"].count().\
                                        sort_values(ascending=False).\
                                        head(10).reset_index()

app_id  install_id
0     174       47416
1     121       39868
2      94       32265
3     189       10082
4      71        8300
5     122        7955
6     154        7694
7      97        7370
8     256        6184
9     104        5286

Application with IDs 174, 94, 121 are our star players!

### Our Top Users

In [263]:
print("Revenue generated by different Top 10 Users\n",
    revenue.groupby("install_id")["value_usd"].sum().sort_values(ascending=False).head(10).reset_index()
     )

Revenue generated by different Top 10 Users
                                           install_id   value_usd
0  b18e32862277cef10fc4f9956671506a888ac19492a0ad...  133.430000
1  ea6b90cc2d2f7e5fcf16eb5b8832111757f1f8f85ab3e7...   95.824843
2  3d5869a49129728cc6086b0d0e9b7d2aa98c4e597e4614...   90.628147
3  3f885b64b8c190673d5c1ee17ffb881c934a6c5cf4c079...   90.413146
4  58386b3403601d1895438ef3fd38279e23db746540bee9...   86.049419
5  c1aa5dda50f680762f24e1961191aee180e3940d4149de...   85.320585
6  f31f1023ac055d4bec20567216c3d2e4251c51fb51c461...   84.515213
7  b04f7c8bc33ef4e6acbbe05f0a2225ba936f50ce9f525d...   83.710000
8  8ef2c9e1f6ef0b07c66a88ec741ad8b2b9ec283838de7e...   82.491656
9  0a2a0fa90cf5dd6ebaa7dd58b23c1d012ead9ad651b90b...   81.530000


### A Journey Through Time

##### Exploring the timeline of our Data..

In [34]:
# Extracting month from event_date
def month(date):
    return date.split("-")[1]

In [36]:
installs["month"] = installs["event_date"].apply(lambda x: month(x))
revenue["month"] = revenue["event_date"].apply(lambda x: month(x))
adspend["month"] = adspend["event_date"].apply(lambda x: month(x))
payouts["month"] = payouts["event_date"].apply(lambda x: month(x))

In [175]:
adspend_month = adspend.groupby("month")["value_usd"].sum()
adspend_month.iplot(kind="line",
                    mode="lines+markers",
                    layout = {
                      "title": "Monthly Ad Expenses in 2022",
                      "xaxis": {"title":"Month"},
                      "yaxis": {"title":"Total Amount Spent"}}
                     )

In [94]:
installs_month = installs.groupby("month")["install_id"].count()
installs_month.iplot(kind="line",
                     mode="lines+markers",
                     layout = {
                         "title": "Monthly Install Trends for 2022",
                         "xaxis": {"title":"Month"},
                         "yaxis": {"title":"Number of Installs"}}
                    )

In [95]:
revenue_month = revenue.groupby("month")["value_usd"].sum()
revenue_month.iplot(kind="line",
                    mode="lines+markers",
                    layout = {
                        "title": "Monthly Revenue for 2022",
                        "xaxis": {"title":"Month"},
                        "yaxis": {"title":"Total Revenue"}
                    }
                   )

In [96]:
payouts_month = payouts.groupby("month")["value_usd"].sum()
payouts_month.iplot(kind="line",
                    mode="lines+markers",
                    layout = {
                        "title": "Monthly Payout for 2022",
                        "xaxis": {"title":"Month"},
                        "yaxis": {"title":"Total Payout"}
                    }
                   )

Let us now see Pearson correlation coefficient, a statistical measure to see how closely the two variables are related.

A little context - Pearson's constant is a measure of the linear relationship between two variables, and ranges from -1 (perfect negative correlation) to 1 (perfect positive correlation), with 0 indicating no correlation.

In [145]:
corr_matrix = pd.concat([adspend_month, installs_month, revenue_month, payouts_month], 
                        axis=1, 
                        keys=['total ad spend', 'num of installs', 'total revenue', 'total payout'],
                       )

In [146]:
corr_matrix.corr().iplot(kind="heatmap",
                        colorscale='Oranges',
                        title= "Tracking Correlations: How Different Metrics Align Across Months")

Most of the metrics represent a strong positive correlation .i.e, increase in the value of one metric leads to the increase ijnj the value of other (as the minimum value of correlation is ~0.77 )

However, the most significant correlation is amongst -

1) <b>adspend and revenue</b>
There is a strong positive correlation(~0.94) between advertising spend and revenue, suggesting that as the expenditure on advertising increases, the revenue generated also increases significantly.

2) <b>payout and number of installs</b>
There appears to be a strong positive correlation(~0.87) between the number of installs and the total payout on the download of partner apps, suggesting that as the number of installs increases, so does the number of partner app downloads



### Final Conclusions

- The countries where our apps are downloaded the most are Country 1 and Country 109.

- The network which gets us the most installs is Network 60.  The second best network is Network 26, where we do not spend any money on advertisments(YET)!<br/>
   In Country 109, Network 26 definitely deserves a mention as it is very close to yielding the  maximum installs

- Applications with IDs 174, 94, 121 are the most expensive to us. But, they are also the most popular amongst users. So, we are doing it RIGHT!

- A time-based analysis showed us that Better Advertisments = Higher Revenue!!<br/>We had three significant months worth noting - April, August and November. 
   Our Advertisement expenditures are very low in the month of April leading to a very small Revenue!<br/>
   August and November are our Top Performers in all ways!<br/>
   Another thing worth noting - More number of Installments does not always mean a Higher Revenue.<br/>

### END -----